specs/kvstore/WriterFencing.fizz

```diff
@@ -0,0 +1,150 @@
+---
+action_options:
+    SlateDBWriter.FlushWal:
+        max_actions: 3
+---
+
+# Model to validate that there can only be one writer and that old writers can be fenced.
+SUCCESS = "success"
+CONFLICT = "conflict"
+NUM_WRITERS = 2
+
+role ObjectStore:
+  action Init:
+    self.objects = {}
+
+  atomic func ConditionalWrite(name, content):
+    if name in self.objects.keys():
+        return CONFLICT
+    self.objects[name] = content
+    return SUCCESS
+
+  atomic func Read(name):
+    return self.objects.get(name)
+
+  atomic func List():
+    return sorted(self.objects.keys())
+
+symmetric role SlateDBWriter:
+    action Init:
+        self.state = "waiting"
+        self.flush_count = 0
+        self.was_ever_open = 0
+    
+    atomic action FlushWal:
+        if self.state != "opened":
+            return
+        self.WriteWal()
+        self.flush_count += 1
+    
+    func LoadLatestManifest():
+        manifests = manifest_store.List()
+        result = CONFLICT
+        if len(manifests) > 0:
+            recent_manifest = manifest_store.Read(manifests[len(manifests) - 1])
+            wal = wal_store.List()
+            self.next_wal = wal[-1] + 1 if wal else 1
+            result = self.WriteManifest(record(id = recent_manifest.id + 1, writer_epoch = recent_manifest.writer_epoch + 1))
+        else:
+            atomic:
+                self.next_wal = 1
+                result = self.WriteManifest(record(id = 1, writer_epoch = 1))        
+        return result
+
+    func WriteManifest(new_manifest):
+        write_status = manifest_store.ConditionalWrite(new_manifest.id, new_manifest)
+        if write_status == SUCCESS:
+            atomic:
+                self.state = "ManifestUpdated"
+                self.manifest = new_manifest
+        else:
+            self.state = "fenced"
+        return write_status
+
+    
+    #db.rs\db\open_with_fp_registry() does below
+    func Open():
+        self.state = "initializing"
+        atomic:
+            load_manifest_result = self.LoadLatestManifest()
+            if load_manifest_result != SUCCESS:
+                return
+        self.WriteFencingWal()
+
+    #flush.rs\DbInner\flush() is the only method that increments next_wal on write.
+    func WriteWal():
+        write_status = wal_store.ConditionalWrite(self.next_wal, self.manifest.writer_epoch)
+        if write_status == SUCCESS:
+            self.next_wal = self.next_wal + 1
+        else:
+            self.state = "fenced"
+        return write_status
+
+    #db.rs\DbInner\fence_writers()
+    func WriteFencingWal():
+        while self.state == "ManifestUpdated":
+            write_status = self.WriteWal()
+            if write_status == SUCCESS:
+                atomic:
+                    self.state = "opened"
+                    self.was_ever_open = 1
+                break
```

Comment:
a minor comment. If the break was within the atomic block, the state space would be smaller.

Ideally this shouldn't have made a difference, but purely because of how the current interpreter is implemented, there will be an yield point before the break statement in this case. So moving the break within the atomic block removes a few states.